pkg/template/template.go

```diff
@@ -19,37 +20,621 @@ package template
+}
+
+// saveIf stores the provided value at specific hierarchy as mentioned in the
+// fields. However, the provided value is not set if the hierarchy i.e. the path
+// determined by the fields has a non-empty value set previously.
+//
+// NOTE:
+//  This hierarchy along with the provided value is formed/updated in the
+// values object.
+//
+// NOTE:
+//  fields is represented as a single string with each field separated by dot
+// i.e. '.'
+//
+// Example:
+// {{- "Hi" | saveIf "TaskResult.msg" .Values | noop -}}
+// {{- "Hi There" | saveIf "TaskResult.msg" .Values | noop -}}
+// {{- .Values.TaskResult.msg -}}
+//
+// Above will print 'Hi'
+//
+// Example:
+// {{- "Hi" | saveIf "TaskResult.msg" .Values | noop -}}
+// {{- "Hi There" | saveAs "TaskResult.msg" .Values | noop -}}
+// {{- .Values.TaskResult.msg -}}
+//
+// Above will print 'Hi There'
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func saveIf(fields string, values map[string]interface{}, value interface{}) interface{} {
+	fieldsArr := strings.Split(fields, ".")
+	oldValue := util.GetNestedField(values, fieldsArr...)
+	// will not override the old value
+	if !empty(oldValue) {
+		return oldValue
+	}
+
+	util.SetNestedField(values, value, fieldsArr...)
+	return value
+}
+
+// notFoundErr throws NotFoundError if given object is empty
+//
+// NOTE:
+//  This function is intended to be used as a go template function
+//
+// Example:
+// {{- "" | notFoundErr "empty object" | saveIf "errMsg" .Values | noop -}}
+//
+// Above returns NotFoundError during template execution. However this
+// does not result in a runtime error.
+//
+// {{- "I am not empty" | notFoundErr "empty object" | saveIf "errMsg" .Values | noop -}}
+//
+// Above does not return any error during template execution.
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func notFoundErr(errMessage string, given interface{}) (err error) {
+	if !empty(given) {
+		// no error if not empty
+		return
+	}
+
+	if len(errMessage) == 0 {
+		errMessage = "item is not found"
+	}
+
+	err = &NotFoundError{
+		err: errMessage,
+	}
+
+	return
+}
+
+// verifyErr returns VerifyError if given verification flag failed i.e. is true
+//
+// NOTE:
+//  This is intended to be used as a go template function
+//
+// Example:
+// {{- "" | empty | verifyErr "empty value provided" | saveIf "errMsg" .Values | noop -}}
+//
+// Above returns VerifyError during template execution. However this
+// does not result in a runtime error.
+//
+// {{- "I am not empty" | empty | verifyErr "empty value provided" | saveIf "errMsg" .Values | noop -}}
+//
+// Above does not return any error during template execution.
+//
+// Assumption here is .Values is of type 'map[string]interface{}'
+func verifyErr(errMessage string, hasVerificationFailed bool) (err error) {
+	if !hasVerificationFailed {
+		// no error if verification did pass successfully
+		return
+	}
+
+	if len(errMessage) == 0 {
+		errMessage = "verification failed"
```

Comment:
if possible we can have the list of error messages in a common location or with typed string.

Reply:
the purpose is different here. this is a template function that accepts error message in case the verification fails. if there are no error message provided then the default message is "verification failed". this default message is only applicable to this particular function.

Reply:
ok